File: backend/app/schemas/jd_analysis.py
from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, HttpUrl


class JDSourceType(str, Enum):
    pasted_text = "pasted_text"
    link = "link"
    screenshot_ocr = "screenshot_ocr"
    manual_note = "manual_note"
    recruiter_message = "recruiter_message"


class RoleType(str, Enum):
    research_engineer = "research_engineer"
    algorithm_engineer = "algorithm_engineer"
    technical_lead = "technical_lead"
    tpm_variant = "tpm_variant"
    business_operation = "business_operation"
    unclear = "unclear"


class MatchLevel(str, Enum):
    high = "high"
    medium_high = "medium_high"
    medium = "medium"
    low = "low"
    not_match = "not_match"


class TimingRecommendation(str, Enum):
    apply_now = "apply_now"
    build_contact_only = "build_contact_only"
    pause_and_prepare = "pause_and_prepare"
    do_not_apply = "do_not_apply"


class ResumeSuggestionStatus(str, Enum):
    pending = "pending"
    accepted = "accepted"
    rejected = "rejected"
    needs_evidence = "needs_evidence"


class JDInput(BaseModel):
    id: str
    source_type: JDSourceType
    company: str = ""
    role_title: str = ""
    location: str = ""
    url: Optional[HttpUrl] = None
    jd_text: str = ""
    recruiter_context: str = ""
    user_question: str = ""
    created_at: datetime


class JDInputCreate(BaseModel):
    source_type: JDSourceType = JDSourceType.pasted_text
    company: str = ""
    role_title: str = ""
    location: str = ""
    url: Optional[HttpUrl] = None
    jd_text: str = ""
    recruiter_context: str = ""
    user_question: str = ""


class JDAnalysis(BaseModel):
    role_type: RoleType
    match_level: MatchLevel
    technical_overlap: List[str] = []
    red_flags: List[str] = []
    fatal_gaps: List[str] = []
    trainable_gaps: List[str] = []
    timing_recommendation: TimingRecommendation
    overall_recommendation: str


class ResumeRevisionSuggestion(BaseModel):
    id: str
    target_section: str
    current_text: str = ""
    suggested_text: str
    reason: str
    risk: str = ""
    status: ResumeSuggestionStatus = ResumeSuggestionStatus.pending


class CapabilityAction(BaseModel):
    id: str
    title: str
    reason: str
    related_gap: str
    suggested_timeframe: str = ""
    status: str = "pending"


class JDAnalysisPayload(BaseModel):
    jd_input: JDInput
    resume_profile_snapshot_id: Optional[str] = None
    learning_plan_snapshot_id: Optional[str] = None
    analysis: JDAnalysis
    resume_revision_suggestions: List[ResumeRevisionSuggestion] = []
    capability_actions: List[CapabilityAction] = []

File: ios/InfraAgent/Models/JDAnalysis.swift
import Foundation

struct JDAnalysisPayload: Codable {
    let jdInput: JDInput
    let resumeProfileSnapshotID: String?
    let learningPlanSnapshotID: String?
    let analysis: JDAnalysis
    let resumeRevisionSuggestions: [ResumeRevisionSuggestion]
    let capabilityActions: [CapabilityAction]

    enum CodingKeys: String, CodingKey {
        case jdInput = "jd_input"
        case resumeProfileSnapshotID = "resume_profile_snapshot_id"
        case learningPlanSnapshotID = "learning_plan_snapshot_id"
        case analysis
        case resumeRevisionSuggestions = "resume_revision_suggestions"
        case capabilityActions = "capability_actions"
    }
}

struct JDInput: Codable, Identifiable {
    let id: String
    let sourceType: String
    let company: String
    let roleTitle: String
    let location: String
    let url: URL?
    let jdText: String
    let recruiterContext: String
    let userQuestion: String
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case sourceType = "source_type"
        case company
        case roleTitle = "role_title"
        case location
        case url
        case jdText = "jd_text"
        case recruiterContext = "recruiter_context"
        case userQuestion = "user_question"
        case createdAt = "created_at"
    }
}

struct JDInputCreate: Codable {
    let sourceType: String
    let company: String
    let roleTitle: String
    let location: String
    let url: URL?
    let jdText: String
    let recruiterContext: String
    let userQuestion: String

    enum CodingKeys: String, CodingKey {
        case sourceType = "source_type"
        case company
        case roleTitle = "role_title"
        case location
        case url
        case jdText = "jd_text"
        case recruiterContext = "recruiter_context"
        case userQuestion = "user_question"
    }
}

struct JDAnalysis: Codable {
    let roleType: String
    let matchLevel: String
    let technicalOverlap: [String]
    let redFlags: [String]
    let fatalGaps: [String]
    let trainableGaps: [String]
    let timingRecommendation: String
    let overallRecommendation: String

    enum CodingKeys: String, CodingKey {
        case roleType = "role_type"
        case matchLevel = "match_level"
        case technicalOverlap = "technical_overlap"
        case redFlags = "red_flags"
        case fatalGaps = "fatal_gaps"
        case trainableGaps = "trainable_gaps"
        case timingRecommendation = "timing_recommendation"
        case overallRecommendation = "overall_recommendation"
    }
}

struct ResumeRevisionSuggestion: Codable, Identifiable {
    let id: String
    let targetSection: String
    let currentText: String
    let suggestedText: String
    let reason: String
    let risk: String
    let status: String

    enum CodingKeys: String, CodingKey {
        case id
        case targetSection = "target_section"
        case currentText = "current_text"
        case suggestedText = "suggested_text"
        case reason
        case risk
        case status
    }
}

struct CapabilityAction: Codable, Identifiable {
    let id: String
    let title: String
    let reason: String
    let relatedGap: String
    let suggestedTimeframe: String
    let status: String

    enum CodingKeys: String, CodingKey {
        case id
        case title
        case reason
        case relatedGap = "related_gap"
        case suggestedTimeframe = "suggested_timeframe"
        case status
    }
}
